# Урок 8

In [29]:
import numpy as np
np.set_printoptions(precision=2, suppress=True)

1. Найти с помощью NumPy SVD для матрицы

$$\begin{pmatrix}
1 & 2 & 0\\ 
0 & 0 & 5\\ 
3 & -4 & 2\\ 
1 & 6 & 5\\ 
0 & 1 & 0
\end{pmatrix}.$$

#### Решение

In [30]:
A = np.array([[1, 2, 0],
              [0, 0, 5],
              [3, -4, 2],
              [1, 6, 5],
              [0, 1, 0]])

U, s, W = np.linalg.svd(A)

# Транспонируем матрицу W
V = W.T

# s - список диагональных элементов, его нужно привести к виду диагональной матрицы для наглядности
D = np.zeros_like(A, dtype=float)
D[np.diag_indices(min(A.shape))] = s

print(f'Матрица U:\n{U}')
print(f'Матрица D:\n{D}')
print(f'Матрица V:\n{V}')

Матрица U:
[[ 0.17  0.16 -0.53 -0.8  -0.16]
 [ 0.39 -0.53  0.61 -0.43  0.03]
 [-0.14 -0.82 -0.52  0.14  0.07]
 [ 0.89  0.06 -0.25  0.38 -0.06]
 [ 0.08  0.11 -0.08 -0.11  0.98]]
Матрица D:
[[8.82 0.   0.  ]
 [0.   6.14 0.  ]
 [0.   0.   2.53]
 [0.   0.   0.  ]
 [0.   0.   0.  ]]
Матрица V:
[[ 0.07 -0.37 -0.93]
 [ 0.72  0.67 -0.21]
 [ 0.69 -0.65  0.31]]


In [31]:
# Проведём проверки
print(np.dot(U.T, U))
print(np.dot(V.T, V))
print(np.dot(np.dot(U, D), V.T))

[[ 1.  0. -0. -0. -0.]
 [ 0.  1.  0. -0.  0.]
 [-0.  0.  1. -0. -0.]
 [-0. -0. -0.  1. -0.]
 [-0.  0. -0. -0.  1.]]
[[ 1.  0.  0.]
 [ 0.  1. -0.]
 [ 0. -0.  1.]]
[[ 1.  2.  0.]
 [ 0. -0.  5.]
 [ 3. -4.  2.]
 [ 1.  6.  5.]
 [-0.  1.  0.]]


2. Для матрицы из предыдущего задания найти:

    а) евклидову норму;
    
    б) норму Фробениуса.

#### Решение

Самый просто способ вычислить евклидову норму - это взять первый элемент матрицы D и она примерно равна 8.82.<br>
Норма Фробениуса: $\sqrt{8.82^2 + 6.14^2 + 2.53^2} \approx 11 $

In [32]:
print(f'а) евклидова норма = {D[0][0]}')
print(f'b) норма Фробениуса = {np.sqrt(np.sum(D**2))}')

а) евклидова норма = 8.824868854820442
b) норма Фробениуса = 11.045361017187261


In [33]:
np.linalg.norm(A)

11.045361017187261

Вывод: в питоне используется норма Фробениуса